plugins/obs-webrtc/whip-output.c

```diff
@@ -0,0 +1,229 @@
+#include "whip-output.h"
+
+static void whip_output_close_unsafe(struct whip_output *output)
+{
+	if (output && output->whip_output) {
+		obs_webrtc_whip_output_close(output->whip_output);
+		obs_webrtc_whip_output_free(output->whip_output);
+		output->whip_output = NULL;
+	}
+}
+
+static const char *whip_output_getname(void *type_data)
+{
+	UNUSED_PARAMETER(type_data);
+
+	return obs_module_text("whip_output");
+}
+
+static void *whip_output_create(obs_data_t *settings, obs_output_t *obs_output)
+{
+	UNUSED_PARAMETER(settings);
+
+	struct whip_output *output = bzalloc(sizeof(struct whip_output));
+	pthread_mutex_init_value(&output->write_mutex);
+
+	output->output = obs_output;
+	output->whip_output = NULL;
+
+	// This needs to be recursive due to `obs_output_signal_stop` calling
+	// `whip_output_total_bytes_sent` (also guarded by mutex)
+	if (pthread_mutex_init_recursive(&output->write_mutex) != 0)
+		goto fail;
+
+	return output;
+
+fail:
+	pthread_mutex_destroy(&output->write_mutex);
+	bfree(output);
+	return NULL;
+}
+
+static void whip_output_destroy(void *data)
+{
+	struct whip_output *output = data;
+
+	pthread_mutex_lock(&output->write_mutex);
+	whip_output_close_unsafe(output);
+	pthread_mutex_unlock(&output->write_mutex);
+
+	pthread_mutex_destroy(&output->write_mutex);
+	bfree(output);
+}
+
+static int map_whip_error_to_obs_output_error(OBSWebRTCWHIPOutputError error)
+{
+	switch (error) {
+	case OBSWebRTCWHIPOutputError_ConnectFailed:
+		return OBS_OUTPUT_CONNECT_FAILED;
+	case OBSWebRTCWHIPOutputError_NetworkError:
+		return OBS_OUTPUT_ERROR;
+	}
```

Comment:
My compiler says:

```
Z:\devel\obs-studio\plugins\obs-webrtc\whip-output.c(62): error C2220: the following warning is treated as an error [Z:\devel\obs-studio\build64\plugins\obs-webrtc\obs-webrtc.vcxproj]
Z:\devel\obs-studio\plugins\obs-webrtc\whip-output.c(62): warning C4715: 'map_whip_error_to_obs_output_error': not all control paths return a value [Z:\devel\obs-studio\build64\plugins\ob
s-webrtc\obs-webrtc.vcxproj]
```

Reply:
I'll take a look!

Reply:
Fixed in 70ee6fbf97f9bd002e3dd066e6921c1f759b8539